src/coreclr/vm/threads.h

```diff
@@ -579,9 +579,25 @@ enum ThreadpoolThreadType
 //
 //---------------------------------------------------------------------------
 Thread* SetupThread();
+Thread* SetupMainThread();
 Thread* SetupThreadNoThrow(HRESULT *phresult = NULL);
-// WARNING : only GC calls this with bRequiresTSL set to FALSE.
-Thread* SetupUnstartedThread(BOOL bRequiresTSL=TRUE);
+
+enum SetupUnstartedThreadFlags
+{
+    SUTF_None = 0,
+
+    // When the thread enters the thread store, the lock should be taken.
```

Comment:
What does it mean for a thread to enter the thread store?

This flag and how it is used does not make a whole lot of sense to me.

Reply:
> What does it mean for a thread to enter the thread store?

Ok, I got confused by some of the places where this is used. I see that it is used in AddThread that the comment is referring to.

Why is it safe to ship the thread store locks for some threads? I understand that it is safe to skip thread store locks early during startup because there is nothing else running at that point. But I do not think it is ever safe to skip the thread store lock to add new threads once the runtime is started up.

Reply:
> What does it mean for a thread to enter the thread store?

The `ThreadStore` class seems to be the keeper of `Thread` instances. Based on some cursory comments it appears to be used to be a singular place to enumerate them. Previously this flag was a `BOOL` passed in. I've not constructed a complete mental model of this code yet so don't have a better answer for why unfortunately. I can say that the only threads that don't currently take this lock are GC threads so I might be able to elide it entirely. It would likely become a specialized `SetupUnstartedThread` instance (e.g., `SetupUnstartedGCThread`).

Reply:
> But I do not think it is ever safe to skip the thread store lock to add new threads once the runtime is started up.

That is true. This flag is only used in places where a lock may be taken - in `ThreadStore::AddThread()` and `Thread::SetBackground()`. Both these functions had a default value of `TRUE` and that was only ever set by the GC  early in a threads life. Based on your statement I am going to say this shouldn't be a high level concept and will instead hide it behind some GC specific idea and have it "just work" for all other scenarios.